parl/algorithms/fluid/sac.py

```diff
@@ -0,0 +1,115 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from parl.core.fluid import layers
+from copy import deepcopy
+import numpy as np
+from paddle import fluid
+from paddle.fluid.layers import Normal
+from parl.core.fluid.algorithm import Algorithm
+
+epsilon = 1e-6
+
+__all__ = ['SAC']
+
+
+class SAC(Algorithm):
+    def __init__(self,
+                 actor,
+                 critic,
+                 action_space,
+                 max_action,
+                 alpha=0.2,
+                 gamma=None,
+                 tau=None,
+                 actor_lr=None,
+                 critic_lr=None):
+        assert isinstance(gamma, float)
+        assert isinstance(tau, float)
+        assert isinstance(actor_lr, float)
+        assert isinstance(critic_lr, float)
+        self.max_action = max_action
+        self.gamma = gamma
+        self.tau = tau
+        self.actor_lr = actor_lr
+        self.critic_lr = critic_lr
+
+        self.alpha = alpha
+
+        self.actor = actor
+        self.critic = critic
+        self.target_critic = deepcopy(critic)
+
+    def predict(self, obs):
+        """ use actor model of self.model to predict the action
+        """
+        mean, _ = self.actor.policy(obs)
+        mean = layers.tanh(mean) * self.max_action
+        return mean
+
+    def sample(self, obs):
+        mean, log_std = self.actor.policy(obs)
+        std = layers.exp(log_std)
+        normal = Normal(mean, std)
+        x_t = normal.sample([1])[0]
+        y_t = layers.tanh(x_t)
+        action = y_t * self.max_action
+        log_prob = normal.log_prob(x_t)
+        log_prob -= layers.log(self.max_action * (1 - layers.pow(y_t, 2)) +
+                               epsilon)
+        log_prob = layers.reduce_sum(log_prob, dim=1, keep_dim=True)
+        log_prob = layers.squeeze(log_prob, axes=[1])
+        return action, log_prob
+
+    def learn(self, obs, action, reward, next_obs, terminal):
+        actor_cost = self.actor_learn(obs)
+        critic_cost = self.critic_learn(obs, action, reward, next_obs,
+                                        terminal)
+        return critic_cost, actor_cost
+
+    def actor_learn(self, obs):
```

Comment:
Both `actor_learn` and `critic_learn` are internal function of SAC,  using `_actor_learn` and `_critic_learn` will be better naming style.